File: CppSandBoxFalseSharing/Execute_2.cpp


/** include
*/
#include "Execute_2.h"
#if(1)


/** Simple
*/
namespace Simple
{
	/** constructor
	*/
	Execute_2::Execute_2(const std::shared_ptr<ShareData>& a_sharedata,int a_index)
	{
		//sharedata
		this->sharedata = a_sharedata;

		//index
		this->index = a_index;
	}

	/** [Simple.WorkThread_Execute_Base]スレッドから呼び出される。
	*/
	void Execute_2::ThreadMain()
	{
		clock_t t_clock = clock();
		{
			int64_t* t_value = &(this->sharedata->value);

			for(int xx=0;xx<ShareData::LOOP_MAX;xx++){
				for(int yy=0;yy<ShareData::LOOP_MAX;yy++){
					*t_value += xx + yy;
				}
				std::atomic_thread_fence(std::memory_order_release);
			}
		}
		std::printf("%d : time = %d\n",this->index,clock() - t_clock);
	}
}


#endif

File: UnitySandBoxFalseSharing/Assets/ReadFromOtherThread/Execute_2.cs


/** ReadFromOtherThread
*/
namespace ReadFromOtherThread
{
	/** Execute_2
	*/
	public sealed class Execute_2 : WorkThread_Execute_Base
	{
		/** mode
		*/
		public int mode;

		/** sharedata
		*/
		public ShareData sharedata;

		/** log
		*/
		public Log log;

		/** constructor
		*/
		public Execute_2(int a_mode,ShareData a_sharedata,Log a_log)
		{
			//mode
			this.mode = a_mode;

			//sharedata
			this.sharedata = a_sharedata;

			//log
			this.log = a_log;
		}

		/** [Simple.WorkThread_Execute_Base]スレッドから呼び出される。
		*/
		public void ThreadMain()
		{
			long t_ticks = System.DateTime.UtcNow.Ticks;
			{
				ref System.UInt64 t_value = ref this.sharedata.value;

				for(int xx=0;xx<ShareData.LOOP_MAX;xx++){
					for(int yy=0;yy<ShareData.LOOP_MAX;yy++){
						t_value += (System.UInt64)(xx + yy);
					}
					System.Threading.Thread.MemoryBarrier();
				}
			}

			lock(this.log){
				this.log.stringbuffer.Append(string.Format("mode = {0} : time = {1}\n",this.mode,System.DateTime.UtcNow.Ticks - t_ticks));
			}
		}
	}
}

